File: SatellaObserver.m
#import "Satella.h"

@implementation SatellaObserver
+ (instancetype) sharedInstance {
	static dispatch_once_t once;
	static id sharedInstance;
	dispatch_once(&once, ^{
		sharedInstance = [[self alloc] init];
	});
	return sharedInstance;
}

- (id) initWithObserver: (id) arg0 {
	self = [super init];
	if (self) self->observer = arg0;
	self->purchases = [[NSMutableArray alloc] init];
	return self;
}

- (void) paymentQueue: (SKPaymentQueue *) queue updatedTransactions:(NSArray<SKPaymentTransaction *> *) transactions {
	for (SKPaymentTransaction *transaction in transactions) {
		
		for (SKPaymentTransaction *purchase in self->purchases) {
			if (purchase.payment == transaction.payment) return;
		}
		
		switch (transaction.transactionState) {
			
			case 1:
				[self->purchases addObject: transaction];
				[queue finishTransaction: transaction];
				break;
				
			case 3:
				if (transaction.originalTransaction) {
					[self->purchases addObject: transaction.originalTransaction];
					[queue finishTransaction: transaction.originalTransaction];
				}
			
			default:
				[transaction _setTransactionState: 1];
				[transaction _setError: nil];
				[self->purchases addObject: transaction];
		}
	}
	
	if (self->purchases) [self->observer paymentQueue: queue updatedTransactions: self->purchases];
}
@end

File: Sources/Satella/SatellaObserver.swift
import SatellaC
import StoreKit

class SatellaObserver: NSObject, SKPaymentTransactionObserver {
	static let shared = SatellaObserver()
	
	public var observer: SKPaymentTransactionObserver? = nil                      // real observer used by the app
	private var purchases                              = [SKPaymentTransaction]() // list of fake purchases made
	
	func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
		for transaction in transactions {                                         // loop through new transactions
			for purchase in purchases {                                           // loop through existing purchases
				if purchase === transaction {		                              // block duplicate transactions by checking memory addresses
					return
				}
			}
			
			switch transaction.transactionState {                                // check each possible state 
				case .purchased:
					purchases.append(transaction)                                // if the transaction is already marked as purchased, add it to 																					the list of purchases
					queue.finishTransaction(transaction)                         // tell the queue to finish processing this transaction
				case .restored:
					if let origTrans = transaction.original {                    // get the original transaction
						purchases.append(origTrans)                              // add it to the list of purchases
						queue.finishTransaction(transaction)                     // mark transaction as having completed
					} else {                                                     // if we can't find the original transaction for whatever reason, 																					just use the default response
						transaction._setTransactionState(.purchased)             // set the code as purchased
						transaction._setError(nil)                               // set the error as nil
						purchases.append(transaction)                            // add to the list
						queue.finishTransaction(transaction)                     // mark transaction as having completed
					}
				default:                                                         // handles all other states with the same code as the else-block 																						in .restored
					transaction._setTransactionState(.purchased)
					transaction._setError(nil)
					purchases.append(transaction)
					queue.finishTransaction(transaction)                         // mark transaction as having completed
			}
		}
		
		observer?.paymentQueue(queue, updatedTransactions: purchases)           // send our hacked list of purchases to the original observer
	}
	
	private override init() {
		super.init()                                                            // this is just overridden to make a proper singleton
	}
}